query.py

```diff
@@ -88,11 +110,51 @@ def __str__(self):
 class QueryAnd(Query):
     def __init__(self, ops):
         super().__init__()
-        self.ops = ops
-        self.ops_size = {} # Dictionary to hold [op: size]
+        self.ops_size = {}  # Dictionary to hold [op: size]
         self.total_size = None
+        self.ops = ops
+        self.ops = self._flatten_ops()
+
+    def _flatten_ops(self):
+        ops = []
+        for op in self.ops:
+            if isinstance(op, QueryAnd):
+                for _op in op._flatten_ops():
+                    ops.append(_op)
+            else:
+                ops.append(op)
 
-    def evaluate(self, inverted_index):
+        return ops
+
+    def evaluate(self, index, **kwargs):
+        return self._evaluate_set(index)
+        # return self._evaluate(index)
+
+    def _evaluate_set(self, inverted_index):
+        add_ops = [set(op.evaluate(inverted_index)) for op in self.ops if not op.is_flipped]
+        negate_ops = [set(op.evaluate(inverted_index)) for op in self.ops if op.is_flipped]
+
+        # case 1, all negate
+        if len(add_ops) == 0:
+            merged = negate_ops[0]
+            for op in negate_ops[1:]:
+                merged.update(op)
+            return list(sorted(inverted_index.all_files.difference(merged)))  # todo can still be improved
+        else:
+            merged = add_ops[0]
+            for op in add_ops[1:]:
+                merged.intersection_update(op)
```

Comment:
Just to double check, after this PR has been merged, i should edit the code here in line 146 to call merge_two_lists function (for skip pointers) instead of intersection_update?

Reply:
yep!